strutil/set.go

```diff
@@ -0,0 +1,99 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package strutil
+
+// OrderedSet is a set of strings that maintains the order of insertion.
+//
+// External synchronization is required for safe concurrent access.
+type OrderedSet struct {
+	items      []string
+	positionOf map[string]int
+}
+
+// Items returns a slice of strings representing insertion order.
+//
+// Contains is O(N) in the size of the set.
+func (o *OrderedSet) Items() []string {
+	return append([]string(nil), o.items...)
+}
+
+// Contains returns true if the set contains a given item.
+//
+// Contains is O(1) in the size of the set.
+func (o *OrderedSet) Contains(item string) bool {
+	if o.positionOf == nil {
+		return false
+	}
+
+	_, ok := o.positionOf[item]
+	return ok
+}
+
+// IndexOf returns the position of an item in the set.
+func (o *OrderedSet) IndexOf(item string) (idx int, ok bool) {
+	idx, ok = o.positionOf[item]
+	return idx, ok
+}
+
+// Del removes an item from the set.
+//
+// Delete is O(N) in the size of the set.
+func (o *OrderedSet) Del(item string) {
+	if o.positionOf == nil {
+		return
+	}
+
+	idx, ok := o.positionOf[item]
+	if !ok {
+		return
+	}
+
+	delete(o.positionOf, item)
+	o.items = append(o.items[:idx], o.items[idx+1:]...)
+
+	if len(o.items) == 0 {
+		o.positionOf = nil
+		o.items = nil
+	}
+}
+
+// Put adds an item into the set.
+//
+// If the item was not present then it is stored and ordered after all existing
+// elements. If the item was already present its position is not changed.
+//
+// Put is O(1) in the size of the set.
+func (o *OrderedSet) Put(item string) {
+	if o.positionOf == nil {
+		o.positionOf = make(map[string]int)
+	}
+
+	if _, ok := o.positionOf[item]; ok {
```

Comment:
This check is only needed if o.positionOf was not null, so could be simplifed to:
```
if o.positionOf == nil {
   ...
} else {
     if _, ok := o.positionOf[item]; ok {
       return
    }
}
```

Reply:
I know and I did code it exactly like so earlier but I think the way I coded it in the end is clearer for the reader.